lib/elixir/pages/Writing Documentation.md

```diff
@@ -70,26 +86,35 @@ When writing documentation:
 
-Elixir treats documentation and code comments as different concepts. Documentation is for users of your Application Programming Interface (API), be it your co-worker or your future self. Modules and functions must always be documented if they are part of your API.
+Code comments are for developers reading the code. They are useful to mark improvements, leave notes for developers reading the code (for example, why you had to resort to a workaround due to a bug in a library) and so forth. Code comments are tied to the source code. You can completely rewrite a function, remove all existing code comments, and it can continue to behave the same, with no change to its behaviour or its documentation.
 
-Code comments are for developers reading the code. They are useful to mark improvements, leave notes for developers reading the code (for example, you decided not to call a function due to a bug in a library) and so forth.
+Because private functions cannot be accessed externally and effectively exist only in your source code, Elixir will warns if a private function has a `@doc` attribute and discards its content. However, you can add code comments to private functions, as with any other piece of code, and we recommend developers to do so whenever they believe it will add relevant information to readers and maintainers of the code.
 
-In other words: documentation is required, code comments are optional.
+Finally, beware of redundant code comments, such as code comments that are describing the exact same as the code:
+
+    # Total is the sum of the batch and individual entries
+    total = batch_sum + individual_sum
+
+In summary, documentation is a contract with users of your API (who do not necessarily have access to source) and code comments are for those who interact directly with the source. You can learn and express different guarantees about your software by separating those two concepts.
 
 ## Hiding Internal Modules and Functions
 
 Besides the modules and functions libraries provide as part of their public interface, libraries may also implement important functionality that is not part of their API. While these modules and functions can be accessed, they are meant to be internal to the library and thus should not have documentation for end users.
 
-Luckily, Elixir allows developers to hide modules and functions from the documentation. For example, one common practice for documenting internal behaviour is to set the `@moduledoc` attribute to `false` while documenting each function:
+Luckily, Elixir allows developers to hide modules and functions from the documentation, by setting `@doc false` to hide a particular function, or `@moduledoc false` to hide the whole module. If a module is hidden, you may even document the functions in the module, but neither the module itself nor any of its functions will be included in the documentation:
```

Comment:
> by setting `@doc false` to hide a particular function, or `@moduledoc false` to hide the whole module. If a module is hidden, you may even document the functions in the module, but neither the module itself nor any of its functions will be included in the documentation:

This is not what actually happens in IEx, where a module with `@moduledoc false` and functions documented with `@doc` can be accessed. Try in IEx `h IEx.History.<TAB>` to see what I mean.